src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -133,20 +136,57 @@ getSMPServerClient c@AgentClient {smpClients, msgQ} srv =
       pure smpVar
 
     waitForSMPClient :: TMVar (Either AgentErrorType SMPClient) -> m SMPClient
-    waitForSMPClient = liftIOEither . atomically . readTMVar
-
+    waitForSMPClient smpVar = do
+      SMPClientConfig {tcpTimeout} <- asks $ smpCfg . config
+      smpClient_ <- liftIO $ tcpTimeout `timeout` atomically (readTMVar smpVar) -- timeout is arbitrary
+      liftEither $ case smpClient_ of
+        Just (Right smpClient) -> Right smpClient
+        Just (Left e) -> Left e
+        Nothing -> Left $ BROKER NETWORK -- TODO ? new BROKER CONNECTING error
+
+    -- TODO ? loop could be inside connectClient function
     newSMPClient :: TMVar (Either AgentErrorType SMPClient) -> m SMPClient
     newSMPClient smpVar =
       tryError connectClient >>= \r -> case r of
         Right smp -> do
           logInfo . decodeUtf8 $ "Agent connected to " <> showServer srv
           atomically $ putTMVar smpVar r
           pure smp
+        Left (BROKER NETWORK) -> do
+          -- same on BROKER TIMEOUT?
+          tryConnectAsync
+          atomically $ do
+            putTMVar smpVar r
+          throwError $ BROKER NETWORK
         Left e -> do
           atomically $ do
             putTMVar smpVar r
             modifyTVar smpClients $ M.delete srv
           throwError e
+      where
+        tryConnectAsync :: m ()
+        tryConnectAsync = do
+          u <- askUnliftIO
+          a <- liftIO $ async . unliftIO u $ connectAsync
```

Comment:
I don't understand why unliftIO is needed here, both tryConnectAsync and connectAsync are in `m` monad